In [32]:
import numpy as np

Sigmoid Perceptron Class

In [33]:
class Sigmoidperceptron():

    def __init__(self,input_size):

        self.weights = np.random.randn(input_size)
        self.bias = np.random.randn(1)
 

    
    def signmoid(self , z):
        return 1/(1+np.exp(-z))



    def predict(self, inputs):
         
         weighted_sum = np.dot(inputs , self.weights) + self.bias

         return self.signmoid(weighted_sum)



    def fit(self, inputs , targets , learning_rate, num_epochs):
        
        num_examples = inputs.shape[0]

        for epoch in range(num_epochs):

            for i in range(num_examples):

                input_vector = inputs[i]

                target = targets[i]

                prediction = self.predict(input_vector)

                error = target - prediction 

                # update weights

                gradient_weights = error * prediction * (1 - prediction) * input_vector
                self.weights += learning_rate * gradient_weights

                # update bias
                gradient_weights = error * prediction * ( 1- prediction)
                self.bias += learning_rate * gradient_weights





    def evaluate(self,inputs , targets):

        correct = 0

        for input_vector , target in zip (inputs ,targets):
            prediction = self.predict(input_vector)

            if prediction  >= 0.5:
                predicted_class = 1

            else:
                predicted_class = 0


            if predicted_class  == target:
                correct+=1

        accuracy = correct / len(inputs)  # accuracy = no. of correct predictions / total number of data points

        return accuracy
        



Importing the dependencies

In [34]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [35]:
df =  pd.read_csv("diabetes_dataset.csv")

In [36]:
df['gender'] = df['gender'].map({'Male': 0, 'Female': 1})

In [37]:
df['gender'] = df['gender'].astype('Int64')

In [38]:
df['smoking_history'] = df['smoking_history'].map({
    'never': 0,
    'current': 1,
    'former': 2,
    'No Info': 3
})

In [39]:
df['smoking_history'] = df['smoking_history'].fillna('10')

In [40]:
df.head()

,gender,age,hypertension,heart_disease,smoking_history,bmi,HbA1c_level,blood_glucose_level,diabetes
0,1,80,0,1,0.0,25.08,6.54,141,0
1,1,55,0,0,3.0,27.33,6.57,80,0
2,0,28,0,0,0.0,27.13,5.74,159,0
3,1,36,0,0,1.0,23.57,5.01,154,0
4,0,75,1,1,1.0,20.24,4.81,156,0


In [41]:
df.tail()

,gender,age,hypertension,heart_disease,smoking_history,bmi,HbA1c_level,blood_glucose_level,diabetes
100987,1,79,0,0,0.0,40.46,8.23,247,1
100988,1,64,0,0,0.0,18.08,5.82,224,1
100989,0,58,0,0,10,36.20,5.70,126,1
100990,1,47,0,0,10,39.86,7.32,186,1
100991,1,43,1,1,10,26.96,6.95,190,1


In [42]:
df.shape

(100992, 9)

In [43]:
df['diabetes'].value_counts()

diabetes
0    92407
1     8585
Name: count, dtype: int64

In [44]:
df = df.drop('gender', axis=1)

In [45]:
class_0_df = df[df['diabetes'] == 0]
class_1_df = df[df['diabetes'] == 1]

class_0_df = class_0_df.sample(8585)

data = pd.concat([class_0_df , class_1_df])

In [46]:
data.isna().sum()

age                    0
hypertension           0
heart_disease          0
smoking_history        0
bmi                    0
HbA1c_level            0
blood_glucose_level    0
diabetes               0
dtype: int64

In [47]:
data.head()

,age,hypertension,heart_disease,smoking_history,bmi,HbA1c_level,blood_glucose_level,diabetes
38397,7,0,0,3.0,15.81,3.48,158,0
23045,77,1,0,0.0,27.05,6.53,158,0
55366,77,0,0,3.0,26.12,4.51,199,0
44486,27,0,0,1.0,27.56,6.65,158,0
31701,1,0,0,3.0,17.79,4.03,156,0


In [48]:
data.shape

(17170, 8)

In [49]:
data['diabetes'].value_counts()

diabetes
0    8585
1    8585
Name: count, dtype: int64

In [50]:
x = data.drop('diabetes', axis=1)

y = data['diabetes']

In [51]:
type(x)

pandas.DataFrame

In [52]:
y

38397     0
23045     0
55366     0
44486     0
31701     0
         ..
100987    1
100988    1
100989    1
100990    1
100991    1
Name: diabetes, Length: 17170, dtype: int64

In [53]:
X = x.values
Y = y.values

In [54]:
type(X)

numpy.ndarray

In [55]:
X

array([[7, 0, 0, ..., 15.81, 3.48, 158],
       [77, 1, 0, ..., 27.05, 6.53, 158],
       [77, 0, 0, ..., 26.12, 4.51, 199],
       ...,
       [58, 0, 0, ..., 36.2, 5.7, 126],
       [47, 0, 0, ..., 39.86, 7.32, 186],
       [43, 1, 1, ..., 26.96, 6.95, 190]], shape=(17170, 7), dtype=object)

In [56]:
Y

array([0, 0, 0, ..., 1, 1, 1], shape=(17170,))

Train Test Split

In [57]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=42)

In [58]:
scaler = StandardScaler()

X_train_scaled = scaler. fit_transform(X_train)

X_test_scaled = scaler. transform(X_test)

In [59]:
X_train_scaled. shape[1]

7

In [60]:
# Sigmoid Perceptron
perceptron = Sigmoidperceptron(input_size=X_train_scaled. shape[1])

In [63]:
# train the pereceptron model
perceptron.fit(inputs=X_train_scaled, targets=Y_train, learning_rate=0.1, num_epochs=100)

Model Evaluation

In [65]:
# evaluate the model for training data
accuracy = perceptron. evaluate(X_train_scaled, Y_train)
accuracy = round(accuracy*100, 2)
print('Training Accuracy =', accuracy,'%')

Training Accuracy = 88.64 %


So, we have implemented the Perceptron using Python

and we have got accuracy of 88.64%